### PR TITLE
Reboot EVAL_CTORS with the new wasm-ctor-eval

### LINKED ISSUE #15402: Support for fast startup by restoring pre-initialized memory
For WASM applications it's theoretically feasible to generate a full heap offline and then do fast application startup by loading that data directly into memory at startup instead of running lots of init code. Is this something that emscripten could be made compatible with? Right now things like C++ constructors and the VFS seem to assume that an application is always being started fresh and there's no obvious way to adjust everything so that this could work.

### PR DESCRIPTION
**Pull request by kripken**
This updates us to use Binaryen's new version of `wasm-ctor-eval`, which can now
do a lot more things, like eval just part of a function, eval to globals, etc. That plus
other changes on the emscripten side that move more things like sbrk into pure
wasm means that we can eval a lot more code.

Previously `-Oz` would enable `EVAL_CTORS`. That was pretty dangerous as
often it does not help code size. You really just need to run with the option and
then measure the code size change vs the startup speed improvement. So this
PR makes us no longer do anything automatically - you must manually build with
`-s EVAL_CTORS`.

A new mode `EVAL_CTORS=2` is also added. This enables `wasm-ctor-eval`'s
new `--ignore-external-input` flag, which ignores the environment, params to
main, etc. This is unsafe, and probably we should have separate options for
these things, but for now this seems useful for experimentation.

Tested by running all of `wasm2` with `EVAL_CTORS=2` enabled, and then
ignoring the failures that are expected (things reading from argv, for example).
Also I ran around 200,000 fuzzer iterations on binaryen.

Example results on `./emcc tests/hello_libcxx.cpp -O3`:

| mode  |  wasm size(bytes) |
|---|---|
| normal  |  136625 |
| EVAL_CTORS=1  |  136616 |
| EVAL_CTORS=2  |  133059 |

The output on the last one is:
```
trying to eval __wasm_call_ctors
  ...success on __wasm_call_ctors.
trying to eval main
  ...partial evalling successful, but stopping since could not eval: call import: wasi_snapshot_preview1.fd_write
  ...stopping
```
It completely evals the ctors, and in main it evals some stuff, until it reaches
a call to print to stdout.

Fixes #15402

### CODE REVIEW

**Comment by sbc100** [comment]
Can you explain a little more about how that safe version (`EVAL_CTORS=1`) might not be desirable by default at some opt levels?   How could it regress anything?  I guess its could increase the size of your static data segment?  Is that the only possible downside?

File: emcc.py [hunk 1]
```diff
@@ -2165,6 +2163,23 @@ def check_memory_setting(setting):
...
+    elif settings.USE_PTHREADS:
+      diagnostics.warning('emcc', 'disabling EVAL_CTORS due to pthreads (passive segments)')
+      settings.EVAL_CTORS = 0
+    elif settings.RELOCATABLE:
+      diagnostics.warning('emcc', 'disabling EVAL_CTORS due to relocatable (movable segments)')
+      settings.EVAL_CTORS = 0
+    elif settings.ASYNCIFY:
+      # In Asyncify exports can be called more than once, and this seems to not
+      # work properly yet (see test_emscripten_scan_registers).
+      diagnostics.warning('emcc', 'disabling EVAL_CTORS due to asyncify')
```
Comment (sbc100) [inline emcc.py:2180]: I all these cases I think its better to just error our given contradictory settings.  That is what we normally do.

**Comment by kripken** [comment]
Yes, that's right (unchanged from before this PR, but we didn't handle it well). It is just very possible that a small amount of code generates a large amount of changes to the data section, which ends up making the wasm potentially a lot bigger. However, that might still be useful in some cases that care more about startup time.

File: tests/test_other.py [hunk 2]
```diff
@@ -6613,136 +6613,74 @@ def test_EM_ASM_i64(self):
...
+    second = test(3000, 1000, 2000, 0xe, 0x8e5) # noqa
+    third  = test(2000, 3000, 1000, 0xe, 0xe58) # noqa
+    print(first, second, third)
+    assert first == second and second == third
+    print('with bad ctor')
+    first  = test(1000, 2000, 3000, 0xf, 0x58f) # noqa; 2 will succeed
+    second = test(3000, 1000, 2000, 0xf, 0x8f5) # noqa; 1 will succedd
+    third  = test(2000, 3000, 1000, 0xf, 0xf58) # noqa; 0 will succeed
+    print(first, second, third)
+    assert first < second and second < third, [first, second, third]
```
Comment (sbc100) [inline tests/test_other.py:6662]:
```
self.assertLess(first, second)
self.assertLess(second, third)
```

File: tools/building.py [hunk 3]
```diff
@@ -660,14 +660,51 @@ def acorn_optimizer(filename, passes, extra_info=None, return_output=False):
...
-def eval_ctors(js_file, binary_file, debug_info=False): # noqa
-  logger.debug('Ctor evalling in the wasm backend is disabled due to https://github.com/emscripten-core/emscripten/issues/9527')
-  return
-  # TODO re-enable
-  # cmd = [PYTHON, path_from_root('tools/ctor_evaller.py'), js_file, binary_file, str(settings.INITIAL_MEMORY), str(settings.TOTAL_STACK), str(settings.GLOBAL_BASE), binaryen_bin, str(int(debug_info))]
-  # if binaryen_bin:
-  #   cmd += get_binaryen_feature_flags()
-  # check_call(cmd)
+def eval_ctors(js_file, wasm_file, debug_info=False): # noqa
+  WASM_CALL_CTORS = '__wasm_call_ctors'
```
Comment (sbc100) [inline tools/building.py:664]: We already have `emscripten.py:WASM_INIT_FUNC` can we unify these?

File: tools/building.py [hunk 4]
```diff
@@ -660,14 +660,51 @@ def acorn_optimizer(filename, passes, extra_info=None, return_output=False):
...
+      args += ['--kept-exports=' + ','.join(kept_ctors)]
+  else:
+    if settings.EXPECT_MAIN:
+      ctor = '_start'
+    else:
+      ctor = '_initialize'
+    args = ['--ctors=' + ctor, '--kept-exports=' + ctor]
+  if settings.EVAL_CTORS == 2:
+    args += ['--ignore-external-input']
+  logger.warning('ctor_evaller: trying to eval global ctors (' + ' '.join(args) + ')')
```
Comment (sbc100) [inline tools/building.py:701]: This isn't really a warning right?  Maybe just `print(..`

Comment (sbc100) [reply tools/building.py:701]: Or `logger.info`

File: tools/building.py [hunk 5]
```diff
@@ -1481,7 +1518,7 @@ def run_binaryen_command(tool, infile, outfile=None, args=[], debug=False, stdou
   cmd += get_binaryen_feature_flags()
   # if we are emitting a source map, every time we load and save the wasm
   # we must tell binaryen to update it
-  if settings.GENERATE_SOURCE_MAP and outfile:
+  if settings.GENERATE_SOURCE_MAP and outfile and tool in ['wasm-opt', 'wasm-emscripten-finalize']:
```
Comment (sbc100) [inline tools/building.py:1521]: Why this?

File: site/source/docs/optimizing/Optimizing-Code.rst [hunk 6]
```diff
@@ -119,6 +119,59 @@ linker can handle a mix wasm object files and LTO object files.  Passing
 Thus, to allow maximal LTO opportunities with the LLVM wasm backend, build all
 source files with ``-flto`` and also link with ``flto``.
 
+EVAL_CTORS
+==========
+
+Building with ``-s EVAL_CTORS`` will evaluate as much code as possible at
```
Comment (sbc100) [inline site/source/docs/optimizing/Optimizing-Code.rst:125]: I think we should skip the space after `-s` these day in our docs.  I know we are not consistent but its more concise and compatible (especially with cmake) so I think we should prefer it.

**Review by sbc100 — APPROVED** [review]

File: tools/building.py (hunk 5 above)
Comment (kripken) [reply tools/building.py:1521]: Only those tools support source maps. `wasm-ctor-eval` for example does not. I'll add a TODO

**Comment by kripken** [comment]
Thanks for the comments, I pushed commits to address them now.

**Review by sbc100 — APPROVED** [review]

File: site/source/docs/optimizing/Optimizing-Code.rst (hunk 6 above)
Comment (Nikolabenakova90) [reply site/source/docs/optimizing/Optimizing-Code.rst]: 607 912 998

**Pull request merged**